src/PatternSet.ts

```diff
@@ -0,0 +1,473 @@
+    }
+
+    /**
+     * Returns all the patterns in the set.
+     */
+    public entries(): Pattern[] {
+        return this.rawPatterns.map(clonePattern);
+    }
+
+    /**
+     * Removes all entries from the pattern set.
+     */
+    public clear() {
+        this.rawPatterns = [];
+        this.compiledPatterns = [];
+    }
+
+    /**
+     * Checks to see if a string matches any of the patterns in this set.
+     * @param str The string to check.
+     * @returns True if matched, false otherwise.
+     */
+    public isMatched(str: string): boolean {
+        for (const pattern of this.compiledPatterns) {
+            if (pattern instanceof RegExp && pattern.test(str)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * Clones the pattern set.
+     */
+    public clone(): PatternSet {
+        return Object.assign(
+            Object.create(Object.getPrototypeOf(this)),
+            {
+                rawPatterns: this.entries(),
+                compilePatterns: this.compiledPatterns.slice(),
+            }
+        )
+    }
+
+    private findIndex(pattern: Pattern): number | undefined {
+        for (let i = 0; i < this.rawPatterns.length; i++) {
+            const patternInSet = this.rawPatterns[i] as (RegexPattern & GlobPattern);
+            if (
+                (pattern as GlobPattern).glob === patternInSet.glob &&
+                (pattern as RegexPattern).regex === patternInSet.regex) {
+                return i;
+            }
+        }
+
+        return undefined;
+    }
+}
+
+/**
+ * Compiles a {@link Pattern} into a regular expression.
+ * This will directly compile regular expression patterns, and convert globs to regex.
+ *
+ * @param pattern The pattern to compile.
+ * @returns The compiled pattern, or an {@link InvalidPattern} if the pattern is not valid.
+ */
+export function compilePattern(pattern: Pattern): RegExp|InvalidPattern {
+    if ('invalid' in pattern) return pattern;
+    if ('glob' in pattern && 'regex' in pattern) {
+        return { invalid: pattern };
+    }
+
+    // Pattern is regex.
+    if ('regex' in pattern) {
+        try {
+            return new RegExp(pattern.regex);
+        } catch (ex) {
+            return {invalid: pattern};
+        }
+    }
+
+    // Pattern is glob.
+    //   '*'     => matches non-directories
+    //   '**'    => matches anything
+    //   '{a,b}' => matches either 'a' or 'b'
+    if ('glob' in pattern) {
+        const globAsRegex = pattern.glob
+            .replace(REGEX_ESCAPE_METACHARS, "\\$&")  // Escape regex metachars.
+            .replace("\\*", `[^${PATH_SEPARATOR}]+`)  // Glob: '*'
+            .replace("\\*\\*", ".*")                             // Glob: '**'
+            .replace(/\\\{([^}]+(?:,[^}]+)+)\\\}/,               // Glob: '{a,b}'
+                (_, [arg]) => `(?:${arg.split(",").join("|")})`)
+
+        return new RegExp(globAsRegex);
```

Comment:
Translating a glob, to RegExp, means that things like `!` can't work. 
Using something like `picomatch` or `micromatch` would be better in my opinion.